lib/rspec-puppet-facts.rb

```diff
@@ -108,19 +97,20 @@ def on_supported_os(opts = {})
       facter_version_filter = RspecPuppetFacts.facter_version_to_filter(facterversion)
       db = FacterDB.get_facts(filter_spec.merge({ :facterversion =>  facter_version_filter }))
 
-      version = facterversion
-      while db.empty? && version !~ /\d+\.0($|\.\d+)/
-        version = RspecPuppetFacts.down_facter_version(version)
-        facter_version_filter = RspecPuppetFacts.facter_version_to_filter(version)
-        db = FacterDB.get_facts(filter_spec.merge({ :facterversion =>  facter_version_filter }))
-      end
-
-      next if db.empty?
-
-      unless version == facterversion
+      if db.empty?
         if RspecPuppetFacts.spec_facts_strict?
           raise ArgumentError, "No facts were found in the FacterDB for Facter v#{facterversion}, aborting"
-        else
+        end
+
+        facterversion_obj = Gem::Version.new(facterversion)
+
+        version = FacterDB.get_facts(filter_spec).map { |facts| Gem::Version.new(facts[:facterversion]) }.sort.reverse.detect { |v| v <= facterversion_obj }
```

Comment:
Does this only get versions that are less than the requested `facterversion`?
https://github.com/mcanevet/rspec-puppet-facts/pull/101 isn't very elegant and I'd prefer something faster, but it does have the advantage of finding the closest match in *both* directions, (older and newer)

Reply:
From the actual code comment above it:
```
    # FacterDB may have newer versions of facter data for which it contains a subset of all possible
    # facter data (see FacterDB 0.5.2 for Facter releases 3.8 and 3.9). In this situation we need to
    # cycle through and downgrade Facter versions per platform type until we find matching Facter data.
```

The fact it previously allowed upgrades was IMHO incorrect. It would allow you to start relying on a Facter 3 fact when you're only on Facter 2 for example.

This is still not perfect, since you might want Facter 5 which dropped legacy facts (in theory), but get downgraded to a Facter 4 which still has them. However, there's a strict setting for that.